### PR TITLE
Centralise fault handling

### PR DESCRIPTION
**Pull request by BenBE**
This should be done as all platforms essentially did the same anyway and there was nothing platform specific.

### CODE REVIEW

File: CRT.c [hunk 1]
```diff
@@ -18,6 +18,11 @@ in the source distribution for its full text.
 #include <string.h>
 #include <locale.h>
 #include <langinfo.h>
```
Comment (cgzones) [inline CRT.c:20]: missing include of `<unistd.h>` for `getpid()`

File: CRT.c [hunk 2]
```diff
@@ -706,3 +716,67 @@ void CRT_setColors(int colorScheme) {
...
    CRT_colors = CRT_colorSchemes[colorScheme];
 }
+
+void CRT_handleSIGSEGV(int signal) {
+   CRT_done();
+
+   fprintf(stderr, "\n\n"
+      "FATAL PROGRAM ERROR DETECTED\n"
+      "============================\n"
+      "Please report this issue at https://htop.dev including the following information:\n"
```
Comment (cgzones) [inline CRT.c:726]: Maybe use https://github.com/htop-dev/htop/issues/new as issue link?

**Review by cgzones — COMMENTED** [review]
FWIW, an example (with sanitizers):
```
FATAL PROGRAM ERROR DETECTED
============================
Please report this issue at https://htop.dev including the following information:

- Your htop version (htop --version)
- Your OS and kernel version (uname -a)
- Your distribution and release (lsb_release -a)
- Likely steps to reproduce (How did it happened?)
- Backtrace of the issue (see below)

Error information:
------------------
A signal 11 was received.

Backtrace information:
----------------------
The following function calls were active when the issue was detected:
---
./htop(backtrace+0x5b)[0x45d8cb]
./htop(CRT_handleSIGSEGV+0xa7)[0x4dc717]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x14140)[0x7f0cbeccd140]
/lib/x86_64-linux-gnu/libc.so.6(_IO_str_overflow+0x121)[0x7f0cbeb4b3a1]
/lib/x86_64-linux-gnu/libc.so.6(_IO_default_xsputn+0x71)[0x7f0cbeb49c61]
/lib/x86_64-linux-gnu/libc.so.6(+0x69e3c)[0x7f0cbeb30e3c]
/lib/x86_64-linux-gnu/libc.so.6(vsprintf+0x84)[0x7f0cbeb3e264]
./htop(vsprintf+0xa7)[0x4451b7]
./htop(sprintf+0xc4)[0x446254]
./htop(EnvScreen_draw+0x21)[0x5401f1]
./htop(InfoScreen_run+0x23b)[0x541eab]
./htop[0x53ea9c]
./htop[0x4dd767]
./htop(ScreenManager_run+0x70d)[0x5145ad]
./htop(main+0x6c0)[0x4e9490]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xea)[0x7f0cbeaedcca]
./htop(_start+0x2a)[0x4267ca]
---

To make the above information more practical to work with,
you should provide a disassembly of your binary.
This can usually be done by running the following command:

   objdump -d -S -w `which htop` > ~/htop.objdump

Please include the generated file in your report.

Running this program with debug symbols or inside a debugger may provide further insights.

Thank you for helping to improve htop!

htop 3.0.2 aborting.

AddressSanitizer:DEADLYSIGNAL
=================================================================
==7330==ERROR: AddressSanitizer: SEGV on unknown address 0x03e800001ca2 (pc 0x7f0cbeb4b3a1 bp 0x000000000074 sp 0x7ffe4f2dde90 T0)
==7330==The signal is caused by a WRITE memory access.
    #0 0x7f0cbeb4b3a1 in _IO_str_overflow libio/strops.c:133:28
    #1 0x7f0cbeb49c60 in __GI__IO_default_xsputn libio/genops.c:399:24
    #2 0x7f0cbeb49c60 in _IO_default_xsputn libio/genops.c:370:1
    #3 0x7f0cbeb30e3b in __vfprintf_internal stdio-common/vfprintf-internal.c:1373:3
    #4 0x7f0cbeb3e263 in __vsprintf_internal libio/iovsprintf.c:96:9
    #5 0x7f0cbeb3e263 in vsprintf libio/iovsprintf.c:105:10
    #6 0x4451b6 in vsprintf (/home/christian/Coding/workspaces/htop/htop+0x4451b6)
    #7 0x446253 in sprintf (/home/christian/Coding/workspaces/htop/htop+0x446253)
    #8 0x5401f0 in EnvScreen_draw /home/christian/Coding/workspaces/htop/EnvScreen.c:36:4
    #9 0x541eaa in InfoScreen_run /home/christian/Coding/workspaces/htop/InfoScreen.c:75:4
    #10 0x53ea9b in actionShowEnvScreen /home/christian/Coding/workspaces/htop/Action.c:527:4
    #11 0x4dd766 in MainPanel_eventHandler /home/christian/Coding/workspaces/htop/MainPanel.c:80:19
    #12 0x5145ac in ScreenManager_run /home/christian/Coding/workspaces/htop/ScreenManager.c:227:19
    #13 0x4e948f in main /home/christian/Coding/workspaces/htop/htop.c:266:4
    #14 0x7f0cbeaedcc9 in __libc_start_main csu/../csu/libc-start.c:308:16
    #15 0x4267c9 in _start (/home/christian/Coding/workspaces/htop/htop+0x4267c9)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV libio/strops.c:133:28 in _IO_str_overflow
==7330==ABORTING
```

File: CRT.c (hunk 2 above)
Comment (BenBE) [reply CRT.c:726]: Or make https://htop.dev/new-issue point there. That's more flexible and allows for us to change where reports drop; even for ancient™ versions.

**Comment by cgzones** [comment]
It seems on FreeBSD `backtrace` requires linking with `-lexecinfo`.
Maybe also do a function-check in `configure.ac` and use `backtrace` only if available.

**Comment by BenBE** [comment]
> It seems on FreeBSD `backtrace` requires linking with `-lexecinfo`.
> [Maybe also do a function-check in `configure.ac` and use `backtrace` only if available.] nvm, its already in a `HAVE_EXECINFO_H` block

Didn't do an in-depth comparison of the `configure.ac` stuff, but comparing the 14 files I removed, there were essentially two versions: The Linux+Backtrace one and the UnsupportedPlatform one. On Darwin (MacOS) the `HAVE_EXECINFO_H` header check was missing for `#include <execinfo.h>`. Apart from this these files were the same.

**Comment by fasterit** [comment]
"Please report this issue at https://htop.dev" -> "Please check whether this issue has already been reported  at https://htop.dev. If the same problem has not been reported before, please create a new issue with the following information:"
(trying to avoid a flood of duplicates if we publish a version that fails for many people)

File: CRT.c [hunk 3]
```diff
@@ -575,9 +578,11 @@ void CRT_restorePrivileges() {
 
 #endif /* HAVE_SETUID_ENABLED */
 
-// TODO: pass an instance of Settings instead.
+static void *backtraceArray[256];
```
Comment (cgzones) [inline CRT.c:581]: Can't this variable be declared in `CRT_handleSIGSEGV()` (without `static`) ?

File: CRT.c [hunk 4]
```diff
@@ -724,3 +735,67 @@ void CRT_setColors(int colorScheme) {
...
+      "Thank you for helping to improve htop!\n"
+      "\n"
+      "htop " VERSION " aborting.\n"
+      "\n"
+   );
+
+   /* Call old sigsegv handler; may be default exit or third party one (e.g. ASAN) */
+   sigaction (signal, &old_sigsegv_handler, NULL);
+
+   kill(getpid(), signal);
```
Comment (cgzones) [inline CRT.c:800]: This can lead to infinite recursion in case the signal handler does not get reset.

File: CRT.c (hunk 3 above)
Comment (BenBE) [reply CRT.c:581]: Potentially yes. I'll have a look, when I'm back near my notebook.

File: CRT.c (hunk 4 above)
Comment (BenBE) [reply CRT.c:800]: Okay, I'll add a check for that case.

**Comment by BenBE** [comment]
I changed the error link to https://htop.dev/issues. I'd suggest pointing this to https://github.com/htop-dev/htop/issues on our server.

File: CRT.c (hunk 4 above)
Comment (cgzones) [reply CRT.c:816]: I was talking about the signals except SIGSEGV:
The SIGSEGV handler is installed with `SA_RESTART` and we are calling the old handler via `sigaction()`.
Installation through signal(2) might behave different on Linux and BSD.
Maybe just call `_exit`?

**Comment by cgzones** [comment]
A further step will be to install a handler for SIGABRT and then call `abort()` insead of `err()` in XAlloc.c, to have stack-traces on Xfunction failures.

File: darwin/DarwinCRT.c [hunk 5]
```diff
@@ -1,35 +0,0 @@
...
-   CRT_done();
-   #ifdef __APPLE__
-   fprintf(stderr, "\n\nhtop " VERSION " aborting. Please report bug at https://htop.dev\n");
-   #ifdef HAVE_EXECINFO_H
-   size_t size = backtrace(backtraceArray, sizeof(backtraceArray) / sizeof(void *));
-   fprintf(stderr, "\n Please include in your report the following backtrace: \n");
-   backtrace_symbols_fd(backtraceArray, size, 2);
-   fprintf(stderr, "\nAdditionally, in order to make the above backtrace useful,");
-   fprintf(stderr, "\nplease also run the following command to generate a disassembly of your binary:");
-   fprintf(stderr, "\n\n   otool -tvV `which htop` > ~/htop.otool");
```
Comment (cgzones) [inline darwin/DarwinCRT.c:26]: ```otool -tvV `which htop` > ~/htop.otool```
maybe print this if `HTOP_DARWIN` is defined? (instead of objdump...)

Comment (BenBE) [reply darwin/DarwinCRT.c:26]: Good catch. :)

File: CRT.c [hunk 6]
```diff
@@ -734,3 +743,88 @@ void CRT_setColors(int colorScheme) {
...
+#ifdef HAVE_EXECINFO_H
+      "- Backtrace of the issue (see below)\n"
+#endif
+      "\n"
+   );
+
+   fprintf(stderr,
+      "Error information:\n"
+      "------------------\n"
+      "A signal %d was received.\n"
```
Comment (cgzones) [inline CRT.c:769]: Maybe also translate with `strsignal(3)`, in case signal have weird numbers on some system?

**Review by cgzones — APPROVED** [review]
LGTM :+1:

xSnprinf failure:
```
FATAL PROGRAM ERROR DETECTED
============================
Please check at https://htop.dev/issues whether this issue has already been reported.
If no similar issue has been reported before, please create a new issue with the following information:

- Your htop version (htop --version)
- Your OS and kernel version (uname -a)
- Your distribution and release (lsb_release -a)
- Likely steps to reproduce (How did it happened?)
- Backtrace of the issue (see below)

Error information:
------------------
A signal 6 was received.

Backtrace information:
----------------------
The following function calls were active when the issue was detected:
---
./htop(backtrace+0x5b)[0x45d8db]
./htop(CRT_handleSIGSEGV+0x175)[0x4eb025]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x14140)[0x7f8bbc8ba140]
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0x141)[0x7f8bbc6efdb1]
/lib/x86_64-linux-gnu/libc.so.6(abort+0x123)[0x7f8bbc6d9537]
./htop(fail+0x1d)[0x54301d]
./htop[0x54343c]
./htop(OpenFilesScreen_scan+0xb1)[0x50d1b1]
./htop(InfoScreen_run+0x1bc)[0x4fd88c]
./htop[0x4d58bc]
./htop[0x501c07]
./htop(ScreenManager_run+0x69b)[0x528edb]
./htop(main+0x6df)[0x4f792f]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xea)[0x7f8bbc6dacca]
./htop(_start+0x2a)[0x4267da]
---

To make the above information more practical to work with,
you should provide a disassembly of your binary.
This can usually be done by running the following command:

   objdump -d -S -w `which htop` > ~/htop.objdump

Please include the generated file in your report.

Running this program with debug symbols or inside a debugger may provide further insights.

Thank you for helping to improve htop!

htop 3.0.3-dev aborting.

Aborted
```

**Review by cgzones — APPROVED** [review]
LGTM :+1:

```
FATAL PROGRAM ERROR DETECTED
============================
Please check at https://htop.dev/issues whether this issue has already been reported.
If no similar issue has been reported before, please create a new issue with the following information:

- Your htop version (htop --version)
- Your OS and kernel version (uname -a)
- Your distribution and release (lsb_release -a)
- Likely steps to reproduce (How did it happened?)
- Backtrace of the issue (see below)

Error information:
------------------
A signal 6 was received.

Backtrace information:
----------------------
The following function calls were active when the issue was detected:
---
./htop(backtrace+0x5b)[0x45d8db]
./htop(CRT_handleSIGSEGV+0x175)[0x4eb025]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x14140)[0x7f8bbc8ba140]
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0x141)[0x7f8bbc6efdb1]
/lib/x86_64-linux-gnu/libc.so.6(abort+0x123)[0x7f8bbc6d9537]
./htop(fail+0x1d)[0x54301d]
./htop[0x54343c]
./htop(OpenFilesScreen_scan+0xb1)[0x50d1b1]
./htop(InfoScreen_run+0x1bc)[0x4fd88c]
./htop[0x4d58bc]
./htop[0x501c07]
./htop(ScreenManager_run+0x69b)[0x528edb]
./htop(main+0x6df)[0x4f792f]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xea)[0x7f8bbc6dacca]
./htop(_start+0x2a)[0x4267da]
---

To make the above information more practical to work with,
you should provide a disassembly of your binary.
This can usually be done by running the following command:

   objdump -d -S -w `which htop` > ~/htop.objdump

Please include the generated file in your report.

Running this program with debug symbols or inside a debugger may provide further insights.

Thank you for helping to improve htop!

htop 3.0.3-dev aborting.

Aborted
```

**Comment by cgzones** [comment]
Thanks, applied.

**Pull request merged**